File: today2/gas-station.ts
/**
 * https://leetcode.com/problems/gas-station/description/
 */
export { canCompleteCircuit };

function canCompleteCircuit(gas: number[], cost: number[]): number {
  for (let i = 0; i < gas.length; i++) {
    if (checkCircuit(gas, cost, i)) return i;
  }

  return -1;
}

function checkCircuit(gas: number[], cost: number[], station: number): boolean {
  let current = station;
  let tank = 0;

  do {
    tank += gas[current];
    tank -= cost[current];
    if (tank < 0) return false;
    current = nextIndex(current, gas.length);
  } while (current !== station);

  return true;
}

function nextIndex(current: number, circuitLength: number) {
  if (current + 1 === circuitLength) return 0;
  else return current + 1;
}

File: returned/python/today11/gas_station.py
# https://leetcode.com/problems/gas-station/
from typing import List

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        total_gas = 0
        for g in gas:
            total_gas += g

        total_cost = 0
        for c in cost:
            total_cost += c
        
        if total_gas - total_cost < 0:
            return -1
        
        for i in range(len(gas) - 1):
            tank = gas[i]
            for j in range(i, len(gas) - 1):
                if tank - cost[j] >= 0:
                    tank = tank - cost[j] + gas[j+1]
                else:
                    break
                if j == len(gas) - 2:
                    return i
        
        return len(gas) - 1